src/samplers/cpu/mod.rs

```diff
@@ -1,173 +1,288 @@
-    let file = File::open(PROC_STAT)
-        .map_err(|e| debug!("could not read {}: {}", PROC_STAT, e))
-        .unwrap();
-    let mut file = BufReader::new(file);
-    parse_proc_stat(&mut file)
+pub fn nanos_per_tick() -> u64 {
+    let ticks_per_second = sysconf::raw::sysconf(sysconf::raw::SysconfVariable::ScClkTck)
+        .expect("Failed to get Clock Ticks per Second") as u64;
+    SECOND / ticks_per_second
 }
 
-fn parse_proc_stat<T: BufRead>(reader: &mut T) -> ProcStat {
-    let mut ret = ProcStat {
-        cpu_total: HashMap::new(),
-    };
-    for line in reader.lines() {
-        let line = line.unwrap();
-        let parts: Vec<&str> = line.split_whitespace().collect();
-        if parts[0] == "cpu" && parts.len() == 11 {
-            ret.cpu_total
-                .insert(Statistic::User, parts[1].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Nice, parts[2].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::System, parts[3].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Idle, parts[4].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Irq, parts[6].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Softirq, parts[7].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Steal, parts[8].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::Guest, parts[9].parse().unwrap_or(0));
-            ret.cpu_total
-                .insert(Statistic::GuestNice, parts[10].parse().unwrap_or(0));
+impl Cpu {
+    pub fn spawn(config: Arc<Config>, metrics: Arc<Metrics<AtomicU32>>, handle: &Handle) {
+        if let Ok(mut cpu) = Cpu::new(config.clone(), metrics) {
+            handle.spawn(async move {
+                loop {
+                    let _ = cpu.sample().await;
+                }
+            });
+        } else if !config.fault_tolerant() {
+            fatal!("failed to initialize cpu sampler");
+        } else {
+            error!("failed to initialize cpu sampler");
         }
     }
-    ret
 }
 
+#[async_trait]
 impl Sampler for Cpu {
-    fn new(
-        config: Arc<Config>,
-        metrics: Metrics<AtomicU32>,
-    ) -> Result<Option<Box<dyn Sampler>>, Error> {
-        if config.cpu().enabled() {
-            Ok(Some(Box::new(Cpu {
-                common: Common::new(config, metrics),
-                nanos_per_tick: crate::common::nanos_per_tick(),
-            }) as Box<dyn Sampler>))
+    type Statistic = CpuStatistic;
+
+    fn new(config: Arc<Config>, metrics: Arc<Metrics<AtomicU32>>) -> Result<Self, failure::Error> {
+        let perf_counters = CHashMap::new();
+        if config.samplers().cpu().perf_events() {
+            #[cfg(feature = "perf")]
+            {
+                let cores = crate::common::hardware_threads().unwrap_or(1024);
```

Comment:
Why 1024 as the default? That seems likely to be confusing.

Reply:
usize max seemed big, but maybe a safer choice. I will think about it more and make a const with justification or set to max

Reply:
What is the case where this will fail? It seems maybe better to just fail completely, but I also don't understand in what situations this might fail.

Reply:
I've changed the logic here to skip perf initialization and log error if it can't determine the number of threads. It would only fail to detect if there's some issue opening or parsing `/sys/devices/system/cpu/present`. This error is fatal if config is set with `fault_tolerant = false` - another case where we'll want to improve on that pattern, but seems to be the best way to handle this